### ISSUE TITLE
Compilation errors after updating targeted version of Roslyn

### ISSUE THREAD

**Issue by thauk-copperleaf** [dotnet/roslyn#25412]
Last year, I used Roslyn 2.4.0 to create a library that would compile an existing (on-disk) solution containing C# projects, load the resulting assembly for one of the projects, then use reflection to do some analysis on the assembly. This was all working fine. My library targeted .NET Standard.

Since coming back to this project 4 months later, changing the library to target .NET Framework 4.7.1, eliminating a dependency on an external assembly, and updating to Roslyn 2.7.0, it seems that my compilation step now fails with a multitude of errors, and I'm trying to figure out why. Keep in mind that the solution passed to my library will compile with no warnings or errors if I were to open it up in an IDE (either Visual Studio or Rider).

The first error is:

```warning CS8021: No value for RuntimeMetadataVersion found. No assembly containing System.Object was found nor was a value for RuntimeMetadataVersion specified through options.```

The fix most people suggest is to add a metadata reference to either typeof(object) or mscorlib, e.g.:

```compilation.AddReferences(MetadataReference.CreateFromFile(typeof(object).GetTypeInfo().Assembly.Location));```

The next group of errors I get is:
    
```error CS0234: The type or namespace name 'Linq' does not exist in the namespace 'System' (are you missing an assembly reference?)```

Similarly, I can fix that with:

```compilation.AddReferences(MetadataReference.CreateFromFile(typeof(System.Linq.Enumerable).GetTypeInfo().Assembly.Location));```

Finally, I get several thousand errors related to the root namespace my company uses, `CL`:

```error CS0246: The type or namespace name 'CL' could not be found (are you missing a using directive or an assembly reference?)```

Questions:

1. Why do I need to add these references when I didn't have to before?

2. Why does compiling the solution with Visual Studio or Rider work, but with Roslyn fail? What are these IDEs doing differently?

3. Related: I do not control the solution I'm compiling, so it's possible that anything could get imported. How can I avoid hardcoding any metadata references at all, which makes my library brittle?

**Comment by agocke** [comment]
@thauk-copperleaf All compilations require references. If your code was working before, it must have provided references. How are you acquiring references right now?

**Comment by thauk-copperleaf** [comment]
@agocke So that means that when VS or Rider compiles that solution, it's providing implicit references? Where are those recorded?

**Comment by agocke** [comment]
Yes, those references are automatically provided by MSBuild, in the case of VS. If you run MSBuild `/t:rebuild` on the command line with `/v:normal` you should be able to invocations of `csc.exe`. Those invocations should have `/r:...` for each of the references passed to the compilation.

**Comment by thauk-copperleaf** [comment]
I'm not sure what was providing references to the compilter before, or how it was working, without me providing those calls to `compilation.AddReferences()`. Did anything change with regards to references, after Roslyn 2.4.0?

**Comment by agocke** [comment]
@thauk-copperleaf No. References have always been required, back to 1.0.0. You can see this in the unit tests, where CreateCompilation always specifies a list of references.

**Comment by thauk-copperleaf** [comment]
Does it have something to do with the fact that my library used to target .NET Core, and I changed it 
 to .NET Framework (created a new csproj, copied the source over, fixed up NuGet package references)? Do references work differently if you're using Roslyn within .NET Core or .NET Framework?

**Comment by thauk-copperleaf** [comment]
After looking at it more closely, the exact same solution I was using last year _is_ compiling. But a newer version of the solution is not. From what I can tell, when I call `project.GetCompilationAsync().Result`, the compilation for the newer csproj is missing every external assembly reference that is listed inside the csproj's XML.

What can cause Roslyn to "lose" these external assembly references, when loading a solution into a workspace?

**Comment by jcouv** [comment]
Tagging @DustinCampbell as this sounds like an identified pitfall with msbuild.

**Comment by jcouv** [comment]
I'm not 100% sure, but the issue you describe sounds like the one [described here](https://github.com/dotnet/roslyn/issues/24308#issuecomment-362362524), including a workaround.
